### ISSUE TITLE
Protection against HW attacks

### ISSUE THREAD

**Issue by dumblob** [nthdimtech/signet-cad#1]
How do you protect the content of the storage (i.e. flash memory) against HW attacks?

Mainly against reading it out e.g. by connecting to appropriate pins on the PCB and messing up with the signals (including clock/crystal). I could not find you're burning some internal transistors to protect against such attacks nor you're allowing the user to lock the flash memory (before any reading or writing can start) by a very long unique key submitted before any use. I'm not sure which locking features the chosen MCU has, but for me this is a very important question.

**Comment by dumblob** [comment]
E.g. https://www.cl.cam.ac.uk/~sps32/mcu_lock.html .

**Comment by nthdimtech** [comment]
Yes this is something that is on my radar. The STM32L series has memory protection features that

1) Prevent the factory boot-loader from begin activated which could allow hostile firmware to be installed
2) Prevent the SWD debugging interface from being activated which could allow flash to be read/written

See the "flash memory protection" section on page 89 of the MCU reference manual: http://www.st.com/resource/en/reference_manual/dm00151940.pdf

I'm planning on enabling this feature as soon as I support firmware update with digital signature checking. Otherwise the firmware update mechanism in the application (not the factory code) is a trivial workaround to the hardware memory protection features.

I'm using an internally generated clock so the only other kind of attack to consider is manipulation of the external voltage levels. The MCU has some protections and alarms for this kind of thing but I haven't investigated them deeply yet. I'll file some more bugs for this.

**Comment by dragon788** [comment]
This probably won't affect the Signet unless it gets smartcard capabilities, but in the vein of security see also the awesome ROCA vulnerability. https://crocs-muni.github.io/roca/
